### ISSUE TITLE
INVALID OUTPUT FORMAT

### ISSUE THREAD

**Issue by jirib** [katapultcloud/ansible_fetch_terraform_backend_outputs#1]
Does it still work?

```
$ /tmp/ansible/hacking/test-module -m ./fetch_terraform_backend_outputs.py -a 'aws_profile=default bucket=web.xxxxx.com object=web_xxxxx_web.tfstate'
* including generated source, if any, saving to: /home/jiri/.ansible_module_generated
* ansiballz module detected; extracted module source to: /home/jiri/debug_dir
***********************************
RAW OUTPUT

Traceback (most recent call last):
  File "/home/jiri/debug_dir/ansible/modules/fetch_terraform_backend_outputs.py", line 164, in <module>
    main()
  File "/home/jiri/debug_dir/ansible/modules/fetch_terraform_backend_outputs.py", line 157, in main
    result = backend_pull(s3_client, module.params)
  File "/home/jiri/debug_dir/ansible/modules/fetch_terraform_backend_outputs.py", line 105, in backend_pull
    return format_data(raw_data)
  File "/home/jiri/debug_dir/ansible/modules/fetch_terraform_backend_outputs.py", line 90, in format_data
    for element in pretty_data['modules']:
KeyError: 'modules'

***********************************
INVALID OUTPUT FORMAT

Traceback (most recent call last):
  File "/tmp/ansible/hacking/test-module", line 238, in runtest
    results = json.loads(out)
  File "/usr/lib/python3.6/json/__init__.py", line 354, in loads
    return _default_decoder.decode(s)
  File "/usr/lib/python3.6/json/decoder.py", line 339, in decode
    obj, end = self.raw_decode(s, idx=_w(s, 0).end())
  File "/usr/lib/python3.6/json/decoder.py", line 357, in raw_decode
    raise JSONDecodeError("Expecting value", s, err.value) from None
json.decoder.JSONDecodeError: Expecting value: line 1 column 1 (char 0)
```

**Comment by gowthamakanthan** [comment]
@jirib  @katapultcloud  How you ran this? Do this is working for you? am getting the below error.

```
ERROR! no action detected in task. This often indicates a misspelled module name, or incorrect module path.

The error appears to have been in '/tmp/test.yml': line 7, column 7, but may
be elsewhere in the file depending on the exact syntax problem.

The offending line appears to be:

  tasks:
    - name: get Terraform networking variables
      ^ here
```

**Comment by MOURIK** [comment]
Hello, 
I think i have the same issue:

Using module file /usr/lib/python3/dist-packages/ansible/modules/system/setup.py
Pipelining is enabled.
<localhost> ESTABLISH LOCAL CONNECTION FOR USER: devops
<localhost> EXEC /bin/sh -c 'AWS_PROFILE=memo /usr/bin/python && sleep 0'
ok: [localhost]
META: ran handlers

TASK [create-tenant : get Terraform networking variables] **********************************************************************************************
task path: /mnt/d/git_repo/mem/id/roles/create-tenant/tasks/main.yml:7
Using module file /home/jj/.ansible/plugins/modules/fetch_terraform_backend_outputs.py
Pipelining is enabled.
<localhost> ESTABLISH LOCAL CONNECTION FOR USER: hakim
<localhost> EXEC /bin/sh -c 'AWS_PROFILE=mem/usr/bin/python && sleep 0'
The full traceback is:
Traceback (most recent call last):
  File "<stdin>", line 102, in <module>
  File "<stdin>", line 94, in _ansiballz_main
  File "<stdin>", line 40, in invoke_module
  File "/usr/lib/python2.7/runpy.py", line 188, in run_module
    fname, loader, pkg_name)
  File "/usr/lib/python2.7/runpy.py", line 82, in _run_module_code
    mod_name, mod_fname, mod_loader, pkg_name)
  File "/usr/lib/python2.7/runpy.py", line 72, in _run_code
    exec code in run_globals
  File "/tmp/ansible_fetch_terraform_backend_outputs_payload_JngDAw/ansible_fetch_terraform_backend_outputs_payload.zip/ansible/modules/fetch_terraform_
backend_outputs.py", line 164, in <module>
  File "/tmp/ansible_fetch_terraform_backend_outputs_payload_JngDAw/ansible_fetch_terraform_backend_outputs_payload.zip/ansible/modules/fetch_terraform_
backend_outputs.py", line 157, in main
  File "/tmp/ansible_fetch_terraform_backend_outputs_payload_JngDAw/ansible_fetch_terraform_backend_outputs_payload.zip/ansible/modules/fetch_terraform_
backend_outputs.py", line 105, in backend_pull
  File "/tmp/ansible_fetch_terraform_backend_outputs_payload_JngDAw/ansible_fetch_terraform_backend_outputs_payload.zip/ansible/modules/fetch_terraform_
backend_outputs.py", line 90, in format_data
KeyError: 'modules'
fatal: [localhost]: FAILED! => {
    "changed": false,
    "module_stderr": "Traceback (most recent call last):\n  File \"<stdin>\", line 102, in <module>\n  File \"<stdin>\", line 94, in _ansiballz_main\n
File \"<stdin>\", line 40, in invoke_module\n  File \"/usr/lib/python2.7/runpy.py\", line 188, in run_module\n    fname, loader, pkg_name)\n  File \"/us
r/lib/python2.7/runpy.py\", line 82, in _run_module_code\n    mod_name, mod_fname, mod_loader, pkg_name)\n  File \"/usr/lib/python2.7/runpy.py\", line 7
2, in _run_code\n    exec code in run_globals\n  File \"/tmp/ansible_fetch_terraform_backend_outputs_payload_JngDAw/ansible_fetch_terraform_backend_outp
uts_payload.zip/ansible/modules/fetch_terraform_backend_outputs.py\", line 164, in <module>\n  File \"/tmp/ansible_fetch_terraform_backend_outputs_paylo
ad_JngDAw/ansible_fetch_terraform_backend_outputs_payload.zip/ansible/modules/fetch_terraform_backend_outputs.py\", line 157, in main\n  File \"/tmp/ans
ible_fetch_terraform_backend_outputs_payload_JngDAw/ansible_fetch_terraform_backend_outputs_payload.zip/ansible/modules/fetch_terraform_backend_outputs.
py\", line 105, in backend_pull\n  File \"/tmp/ansible_fetch_terraform_backend_outputs_payload_JngDAw/ansible_fetch_terraform_backend_outputs_payload.zi
p/ansible/modules/fetch_terraform_backend_outputs.py\", line 90, in format_data\nKeyError: 'modules'\n",
    "module_stdout": "",
    "msg": "MODULE FAILURE\nSee stdout/stderr for the exact error",
    "rc": 1
}

**Comment by ideologysec** [comment]
The issue is that the format of the terraform state file has changed. Here's a fix that worked for me with Terraform 0.14 (state file version 4). Replace the `format_data` function with the below (also relies on python3, if you use python2, change `permanent.items` to `permanent.iteritems`:

```python
def format_data(data):
    pretty_data = json.loads(data)
    result = {}
    permanent = pretty_data['outputs']

    for key, value in permanent.items():
        result[key] = value['value']

    return result
```

cc @katapultcloud @MOURIK @gowthamakanthan @jirib

**Comment by katapultcloud** [comment]
@ideologysec @MOURIK @gowthamakanthan @jirib 
hey guys thanks for notifying me about this issue. My apologies that I ignored it. Special thanks to @ideologysec for the investigation and fix. I've merged the fix which works :) https://github.com/katapultcloud/ansible_fetch_terraform_backend_outputs/pull/2
